### PR TITLE
Upgrade libraries (in particular winit)

### LINKED ISSUE #145: Alt+Tab on Windows is ineffective
As was mentionned at #144:

AltTabbing while in fullscreen, basically doesn't work. Even if window switcher shows up (it sometimes doesn't) some weird window resizing happens (like it is exiting fullscreen, but it isn't) and chosen window is still beneath.

### LINKED ISSUE #176: crash report
```


--
buffering RedrawRequested event
--

Location C:\Users\runneradmin\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.22.0\src\platform_impl\windows\event_loop\runner.rs:110:17

   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>
   8: <unknown>
   9: DefSubclassProc
  10: DefSubclassProc
  11: CallWindowProcW
  12: CallWindowProcW
  13: glPushClientAttrib
  14: CallWindowProcW
  15: DispatchMessageW
  16: IsWindowVisible
  17: KiUserCallbackDispatcher
  18: NtUserDispatchMessage
  19: DispatchMessageW
  20: InputFocusMonitorCreate
  21: InputFocusMonitorCreate
  22: TF_Notify
  23: TF_Notify
  24: TF_SendLangBandMsg
  25: <unknown>
  26: DllRegisterServer
  27: TF_SendLangBandMsg
  28: TF_SendLangBandMsg
  29: TF_SetDefaultRemoteKeyboardLayout
  30: SetInputScopes2
  31: CtfImeProcessCicHotkey
  32: CtfImeProcessCicHotkey
  33: CtfImeAssociateFocus
  34: CtfImeAssociateFocus
  35: CtfImeAssociateFocus
  36: ImmSetActiveContext
  37: GetWindow
  38: GetWindowRgnBox
  39: GetDlgItem
  40: GetDlgItem
  41: CallWindowProcW
  42: DispatchMessageW
  43: IsWindowVisible
  44: KiUserCallbackDispatcher
  45: NtUserMessageCall
  46: GetWindowTextW
  47: GetWindowTextW
  48: CallWindowProcW
  49: CallWindowProcW
  50: DefSubclassProc
  51: DefSubclassProc
  52: DefSubclassProc
  53: DefSubclassProc
  54: CallWindowProcW
  55: CallWindowProcW
  56: glPushClientAttrib
  57: CallWindowProcW
  58: DispatchMessageW
  59: IsWindowVisible
  60: KiUserCallbackDispatcher
  61: NtUserShowWindow
  62: <unknown>
  63: <unknown>
  64: <unknown>
  65: <unknown>
  66: <unknown>
  67: <unknown>
  68: <unknown>
  69: BaseThreadInitThunk
  70: RtlUserThreadStart

===================================================================================================

--
buffering RedrawRequested event
--

Location C:\Users\runneradmin\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.22.0\src\platform_impl\windows\event_loop\runner.rs:110:17

   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>
   8: <unknown>
   9: DefSubclassProc
  10: DefSubclassProc
  11: CallWindowProcW
  12: CallWindowProcW
  13: glPushClientAttrib
  14: CallWindowProcW
  15: DispatchMessageW
  16: IsWindowVisible
  17: KiUserCallbackDispatcher
  18: NtUserDispatchMessage
  19: DispatchMessageW
  20: InputFocusMonitorCreate
  21: InputFocusMonitorCreate
  22: TF_Notify
  23: TF_Notify
  24: TF_SendLangBandMsg
  25: <unknown>
  26: DllRegisterServer
  27: TF_SendLangBandMsg
  28: TF_SendLangBandMsg
  29: TF_SetDefaultRemoteKeyboardLayout
  30: SetInputScopes2
  31: CtfImeProcessCicHotkey
  32: CtfImeProcessCicHotkey
  33: CtfImeAssociateFocus
  34: CtfImeAssociateFocus
  35: CtfImeAssociateFocus
  36: ImmSetActiveContext
  37: GetWindow
  38: GetWindowRgnBox
  39: GetDlgItem
  40: GetDlgItem
  41: CallWindowProcW
  42: DispatchMessageW
  43: IsWindowVisible
  44: KiUserCallbackDispatcher
  45: NtUserMessageCall
  46: GetWindowTextW
  47: GetWindowTextW
  48: CallWindowProcW
  49: CallWindowProcW
  50: DefSubclassProc
  51: DefSubclassProc
  52: DefSubclassProc
  53: DefSubclassProc
  54: CallWindowProcW
  55: CallWindowProcW
  56: glPushClientAttrib
  57: CallWindowProcW
  58: DispatchMessageW
  59: IsWindowVisible
  60: KiUserCallbackDispatcher
  61: NtUserShowWindow
  62: <unknown>
  63: <unknown>
  64: <unknown>
  65: <unknown>
  66: <unknown>
  67: <unknown>
  68: <unknown>
  69: BaseThreadInitThunk
  70: RtlUserThreadStart

===================================================================================================
```

### PR DESCRIPTION
**Pull request by vn971**
Fixes #176
Fixes #145

### CODE REVIEW

File: subcrates/gelatin/Cargo.toml [hunk 1]
```diff
@@ -11,13 +11,8 @@ edition = "2018"
...
-cgmath = "0.17"
-image = "0.23.6"
-derive_builder = "0.9"
-glium = { version = "0.27" }
-
-# Forcing 0.22.0 to avoid running into
-# https://github.com/rust-windowing/winit/issues/1587
-# The next release that will include the fix will be 
-# AFTER 0.22.2
-winit = "=0.22.0"
```
Comment (vn971) [inline subcrates/gelatin/Cargo.toml:23]: Note that winit is upgraded and the warning is thus removed

File: Cargo.toml [hunk 2]
```diff
@@ -59,32 +42,49 @@ linux_mime_types = [
...
+clap = { version = "2.33.3", default-features = false }
+kamadak-exif = "0.5.3"
+arboard = "1.1.0"
+resvg = "0.13.1"
+usvg = "0.13.1"
+tiny-skia = "0.4.2"
 
 [dependencies.libavif-image]
-version = "0.5"
+version = "0.5.0"
```
Comment (vn971) [inline Cargo.toml:87]: These and other Cargo* changes are the result of running `cargo upgrade`, `cargo update` and then fixing the compilation issues

Comment (kurnevsky) [reply Cargo.toml:87]: Probably better not to specify minor versions? They are locked with Cargo.lock anyway.

Comment (vn971) [reply Cargo.toml:87]: @kurnevsky this is what `cargo upgrade` did by default, I tried not to interfere with manual overrides...

**Comment by vn971** [comment]
The CI errors are about some code that I didn't touch. I could try fix that though. Should I? (CC @ArturKovacs)

**Comment by ArturKovacs** [comment]
If am pretty sure, you are getting these errors because you updated the `ureq` dependency to a new major version (1 -> 2), which unsuprisingly has a different API. So yes, you should fix these.

To see these errors on your own computer, you should enable the networking feature when compiling locally. (Preferably you should enable all features but avif is tricky to set up). You can do this by something like
```
cargo check --features=networking
```

**Comment by benediktwerner** [comment]
There's a bit of a mismatch with the keys going on now. `Plus` and `Minus` get mapped to `Add` and `Subtract` but `Asterisk` doesn't become `Multiply`.

I'm wondering, why are the first two even combined in the first place? It seems sensible to separate them. If you want both to trigger something you can just bind both to a key but if you want to differentiate between them you can't with the current system.

**Comment by vn971** [comment]
Hey, I've fixed the code around `ureq`. I don't know what to do with the AtomicU32 deprecation warning though -- could anybody help? It's deprecated in Rust itself since "1.50.0"

**Comment by benediktwerner** [comment]
The solution is to replace it with `compare_exchange`. The [docs](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU32.html#method.compare_and_swap) explain what the new orderings should be. But I think that can be done in a separate PR. I already offered to fix this and a few of the other clippy warnings in another PR but I haven't heard from Artúr yet.

**Comment by vn971** [comment]
@benediktwerner indeed I tried to use `compare_exchange`, but then the result is a `Result`, so I had to either retry or return the Result out. I had it timeboxed and decided to abort in case somebody knows what to do better.

Let's wait for Artúr's then

**Comment by benediktwerner** [comment]
The operation could already "fail" before. It's just that `compare_and_swap` signals failure by returning something that is different from the passed in `current` value and you had to check it yourself. But in both cases, a failure doesn't really mean that something went wrong. It just means that the `current` value was different and nothing was changed. But this could be completely expected. Since the current use of `compare_and_sawp` ignores the returned value the equivalent code could just discard the `Result`.

**Comment by ArturKovacs** [comment]
Yeah replace those with `compare_exchange` and (I guess) use `Ordering::SeqCst` for both success and failure. We should ignore the result of the function which can be done by
```rust
let _ = PRIORITY_REQUEST_ID.compare_exchange(...);
```

> I already offered to fix this and a few of the other clippy warnings in another PR but I haven't heard from Artúr yet

We should probably merge this before #182

**Comment by vn971** [comment]
Thanks @benediktwerner @ArturKovacs for the explanations! Fixed that warning now, let's see if the CI will succeed.

BTW, I'm not sure if it's checked in CI, but I also see a lot of (unrelated) clippy warning across the codebase. I wonder if it's something that emulsion as a project would want to do (to fix the warnings and enforce lack thereof in CI)

**Comment by vn971** [comment]
Maybe it's the newer version of clippy that finds more inconsistencies tho

**Comment by vn971** [comment]
Maybe it's my local clippy that has a newer version and complains. CI just succeeded

**Comment by vn971** [comment]
I've created a PR for clippy/CI: https://github.com/ArturKovacs/emulsion/pull/184

File: src/utils.rs [hunk 3]
```diff
@@ -204,19 +203,29 @@ pub fn virtual_keycode_to_string(vk: VirtualKeyCode) -> String {
 		// Alphanumeric is `Minus`, the numapd is `Subtract`
```
Comment (ArturKovacs) [inline src/utils.rs:203]: The comment here should say
```
// Alphanumeric is `Minus`, the numapd is `NumpadSubtract`
```

**Comment by ArturKovacs** [comment]
> There's a bit of a mismatch with the keys going on now. `Plus` and `Minus` get mapped to `Add` and `Subtract` but `Asterisk` doesn't become `Multiply`.
> 
> I'm wondering, why are the first two even combined in the first place? It seems sensible to separate them. If you want both to trigger something you can just bind both to a key but if you want to differentiate between them you can't with the current system.

This follows the way it's been in emulsion, so this is correct for now. This is due to a limitation in winit's current keyboard input API. When you press a key which generates a character, you get both a `KeyInput` and a `ReceivedCharacter`. The generated unicode character is *not* known in the `KeyInput` event, and there's no reliable way to tell which `ReceivedCharacter` belongs to which `KeyInput`. But I wanted to allow users to specify unicode characters for key-bindings, instead of having to figure out which virtual keycode they should specify in the config. So the way emulsion handles this, is that whenever it receives a `KeyInput`, it checks whether the pressed key would generate a character, and if it does, then we don't handle the input because expect that the key-binding will be triggered through `ReceivedCharacter`.

File: src/utils.rs [hunk 4]
```diff
@@ -52,26 +52,27 @@ pub fn virtual_keycode_is_char(vk: VirtualKeyCode) -> bool {
 		VirtualKeyCode::Numpad9 => true,
 		VirtualKeyCode::AbntC1 => true,
 		VirtualKeyCode::AbntC2 => true,
-		VirtualKeyCode::Add => true,
 		VirtualKeyCode::Apostrophe => true,
 		VirtualKeyCode::At => true,
```
Comment (ArturKovacs) [inline src/utils.rs:56]: Add the missing `Asterisk` variant here (which is '*'). This actually shouldn't make a difference in how emulsion works, but let's keep things consistent.

**Review by ArturKovacs — CHANGES_REQUESTED** [review]

File: src/utils.rs (hunk 3 above)
Comment (vn971) [reply src/utils.rs]: :+1: done

File: src/utils.rs (hunk 4 above)
Comment (vn971) [reply src/utils.rs]: Done

**Comment by ArturKovacs** [comment]
Thanks a lot!

**Pull request merged**